DigitalLearningSolutions.Data.Tests/Services/UserDataServiceTests.cs

```diff
@@ -0,0 +1,76 @@
+﻿namespace DigitalLearningSolutions.Data.Tests.Services
+{
+    using System.Linq;
+    using DigitalLearningSolutions.Data.Services;
+    using DigitalLearningSolutions.Data.Tests.Helpers;
+    using FluentAssertions;
+    using NUnit.Framework;
+
+    public class UserDataServiceTests
+    {
+        private IUserDataService userDataService;
+
+        [SetUp]
+        public void Setup()
+        {
+            var connection = ServiceTestHelper.GetDatabaseConnection();
+            userDataService = new UserDataService(connection);
+        }
+
+        [Test]
+        public void GetAdminUserByUsername_Returns_admin_user()
+        {
+            // Given
+            var expectedAdminUser = UserTestHelper.GetDefaultAdminUser();
+
+            //When
+            var returnedAdminUser = userDataService.GetAdminUserByUsername("Username");
+            returnedAdminUser.Id = expectedAdminUser.Id;
+
+            // Then
+            returnedAdminUser.Should().BeEquivalentTo(expectedAdminUser);
+        }
+
+        [Test]
+        public void GetDelegateUsersByUsername_Returns_delegate_users()
+        {
+            // Given
+            var expectedDelegateUsers = UserTestHelper.GetDefaultDelegateUser();
+
+            //When
+            var returnedDelegateUsers = userDataService.GetDelegateUsersByUsername("SV1234");
+            returnedDelegateUsers.First().Id = expectedDelegateUsers.Id;
+
+            // Then
+            returnedDelegateUsers.FirstOrDefault().Should().BeEquivalentTo(expectedDelegateUsers);
+        }
+
+        [Test]
+        public void GetAdminUserById_Returns_admin_user()
+        {
+            // Given
+            var expectedAdminUser = UserTestHelper.GetDefaultAdminUser();
+
+            // When
+            var returnedAdminUser = userDataService.GetAdminUserById(4259);
+            returnedAdminUser.Id = expectedAdminUser.Id;
+
+            // Then
+            returnedAdminUser.Should().BeEquivalentTo(expectedAdminUser);
+        }
+
+        [Test]
+        public void GetDelegateUserById_Returns_delegate_users()
+        {
+            // Given
+            var expectedDelegateUsers = UserTestHelper.GetDefaultDelegateUser();
+
+            //When
+            var returnedDelegateUser = userDataService.GetDelegateUserById(299639);
```

Comment:
I'm a little bit nervous doing this by id because it will most likely break for some people locally (as the add test users script can be ran on the db backup at different times). At the same time, I don't want any password hashes being hardcoded in the codebase. 
Do you think we could potentially fetch the users based on an id on the db backup (e.g. 3, which will be the same for everyone) and then check that the fields except for the password hash are as expected? If doing it is a bit messy, we can just check that the fields we care about are as expected (the ones you'll be using on the My account page.

Reply:
Or to be honest, we could just create a script that replaces one of the user password hashes that is in the db backup with "password" and then compare against that (which I think is even better than creating/using the new test users) 👀  ?

Reply:
Any ideas/opinions? :)

Reply:
Yeah didn't think about whether it would break on other people's environments. It would definitely be messy (I guess that is to be expected as these aren't really unit tests) but I could write tests that query against known users from the db backup. Would end up hardcoding in a load of expected user values with the details I need to check instead.

Reply:
I don't believe any password hashes are currently stored in the codebase. The UserTestHelper defaults just have "password" as the password by default, and I don't think that is overwritten anywhere. The test user that that is based on also just has password as "password". If one of the records in the db backup (prior to migrations/scripts), can be safely modified to be the new test user, I don't see any problem with, and is almost certainly preferable to the new test user with the potentially varying Id.

Reply:
Cool - so maybe we can replace the test user script with a different script that just changes the pw of a single candidate and single admin record that are both present in the db backup (and maybe modify some other fields we might need)? Then, we can make default users have these fields and use the same users for all of these data tests.

Reply:
I've updated the PopulateTestUsers script to instead update the first record (other than the admin record ID 1) of each table with the properties we were setting on the created users. These are AdminID 7 and CandidateID 1.
I've also added an update statement for the CentreID 2 so we can set a few of the properties on there as needed.
We should now be able to test against these users/centre for all of the tests we need to, amending the update statements with further data as necessary (e.g. I'll be adding values to the AnswerX fields for 392)

Reply:
This may mean we need to remove the created test users from any databases where they've been added